### PR TITLE
Merge `cdat-migration-fy24` branch into `main`

### LINKED ISSUE #797: [Bug]: Silent bug in `adjust_prs_val_units()` conditional where `if prs_val0:` will be `False` if `prs_val0` is 0
#### Overview
There is a conditional in `adjust_prs_val_units()` that checks `if prs_val0:` to perform unit adjustment on the prs axis.  This is a silent bug in this conditional where unit adjustment is not being performed when `prs_val0 == 0` because `0 == False` in boolean form.

Related code: https://github.com/E3SM-Project/e3sm_diags/blob/43f5f74057929c4bedddfab58be63f838ee36abe/e3sm_diags/derivations/acme.py#L433-L446

#### Example Case
There are several derived variables that define `prs_low0=0`. For example, notice how `CLDLOW_TAU_1.3_9.4_MISR` has a `prs_low0=0` below. For the dataset I was working with, this value is never used due to the silent bug. No prs axis adjustments occur and the prs low value from the dataset (-0.001) is used which results in incorrect results.
 
https://github.com/E3SM-Project/e3sm_diags/blob/43f5f74057929c4bedddfab58be63f838ee36abe/e3sm_diags/derivations/acme.py#L1378-L1393

### What did you expect to happen? Are there are possible answers you came across?

The conditional should be updated to `if prs_val0 is not None:` since `prs_val0` is optional. This logic is implemented correctly on the dev branch, `cdat-migration-fy24`.

```python
def adjust_prs_val_units(
    prs: "FileAxis", prs_val: float, prs_val0: Optional[float]
) -> float:
    """Adjust the prs_val units based on the prs.id"""
    # COSP v2 cosp_pr in units Pa instead of hPa as in v1
    # COSP v2 cosp_htmisr in units m instead of km as in v1
    adjust_ids = {"cosp_prs": 100, "cosp_htmisr": 1000}


    if prs_val0 is not None:
        prs_val = prs_val0
        if prs.id in adjust_ids.keys() and max(prs.getData()) > 1000:
            prs_val = prs_val * adjust_ids[prs.id]


    return prs_val
```

### Minimal Complete Verifiable Example (MVCE)

_No response_

### Relevant log output

_No response_

#### How I found it

In PR #794, I found the following variables had large diffs. I stepped through the code on `main` and the dev branch and found that `main` doesn't perform prs unit adjustment while the dev branch does.
```bash
Comparing:
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/792-lat-lon/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU1.3_9.4_MISR-ANN-global_ref.nc
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU1.3_9.4_MISR-ANN-global_ref.nc
    * var_key: CLDLOW_TAU1.3_9.4_MISR

Not equal to tolerance rtol=1e-05, atol=0

Mismatched elements: 39457 / 64800 (60.9%)
Max absolute difference: 22.411116
Max relative difference: 0.6832267
 x: array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],...
 y: array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],...
Comparing:
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/792-lat-lon/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU1.3_9.4_MISR-ANN-global_test.nc
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU1.3_9.4_MISR-ANN-global_test.nc
    * var_key: CLDLOW_TAU1.3_9.4_MISR

Not equal to tolerance rtol=1e-05, atol=0

Mismatched elements: 9 / 64800 (0.0139%)
Max absolute difference: 0.0970192
Max relative difference: 0.01244658
 x: array([[ 0.056777,  0.056777,  0.056777, ...,  1.274017,  1.274017,
         1.274017],
       [ 0.207892,  0.207774,  0.207536, ...,  1.675944,  1.676576,...
 y: array([[ 0.056777,  0.056777,  0.056777, ...,  1.274017,  1.274017,
         1.274017],
       [ 0.207892,  0.207774,  0.207536, ...,  1.675944,  1.676576,...
Comparing:
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/792-lat-lon/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU1.3_9.4_MISR-JJA-global_ref.nc
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU1.3_9.4_MISR-JJA-global_ref.nc
    * var_key: CLDLOW_TAU1.3_9.4_MISR

Not equal to tolerance rtol=1e-05, atol=0

Mismatched elements: 34699 / 64800 (53.5%)
Max absolute difference: 45.429226
Max relative difference: 0.9708206
 x: array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],...
 y: array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],...

------------------------------------------------------------------------------------------------------------------------------------------------

Comparing:
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/792-lat-lon/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU1.3_MISR-ANN-global_ref.nc
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU1.3_MISR-ANN-global_ref.nc
    * var_key: CLDLOW_TAU1.3_MISR

Not equal to tolerance rtol=1e-05, atol=0

Mismatched elements: 39499 / 64800 (61%)
Max absolute difference: 37.673122
Max relative difference: 0.62295455
 x: array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],...
 y: array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],...
Comparing:
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/792-lat-lon/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU1.3_MISR-ANN-global_test.nc
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU1.3_MISR-ANN-global_test.nc
    * var_key: CLDLOW_TAU1.3_MISR

Not equal to tolerance rtol=1e-05, atol=0

Mismatched elements: 9 / 64800 (0.0139%)
Max absolute difference: 0.0970192
Max relative difference: 0.00541773
 x: array([[5.677656e-02, 5.677656e-02, 5.677656e-02, ..., 1.274017e+00,
        1.274017e+00, 1.274017e+00],
       [2.078919e-01, 2.077735e-01, 2.075364e-01, ..., 1.675944e+00,...
 y: array([[5.677656e-02, 5.677656e-02, 5.677656e-02, ..., 1.274017e+00,
        1.274017e+00, 1.274017e+00],
       [2.078919e-01, 2.077735e-01, 2.075364e-01, ..., 1.675944e+00,...
Comparing:
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/792-lat-lon/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU1.3_MISR-JJA-global_ref.nc
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU1.3_MISR-JJA-global_ref.nc
    * var_key: CLDLOW_TAU1.3_MISR

Not equal to tolerance rtol=1e-05, atol=0

Mismatched elements: 35149 / 64800 (54.2%)
Max absolute difference: 67.89603
Max relative difference: 0.9691263
 x: array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],...
 y: array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],...

------------------------------------------------------------------------------------------------------------------------------------------------

Comparing:
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/792-lat-lon/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU9.4_MISR-ANN-global_ref.nc
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU9.4_MISR-ANN-global_ref.nc
    * var_key: CLDLOW_TAU9.4_MISR

Not equal to tolerance rtol=1e-05, atol=0

Mismatched elements: 39323 / 64800 (60.7%)
Max absolute difference: 31.085188
Max relative difference: 0.96666664
 x: array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],...
 y: array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],...

Comparing:
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/792-lat-lon/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU9.4_MISR-JJA-global_ref.nc
    * /global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main/lat_lon/Cloud MISR/MISRCOSP-CLDLOW_TAU9.4_MISR-JJA-global_ref.nc
    * var_key: CLDLOW_TAU9.4_MISR

Not equal to tolerance rtol=1e-05, atol=0

Mismatched elements: 33486 / 64800 (51.7%)
Max absolute difference: 63.126827
Max relative difference: 1.
 x: array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],...
 y: array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],...

```

#### Affects the following derived variables:
https://github.com/E3SM-Project/e3sm_diags/blob/43f5f74057929c4bedddfab58be63f838ee36abe/e3sm_diags/derivations/acme.py#L1362-L1409

### Environment

`main` and all versions of `e3sm_diags` that has `cosp_bin_sum()`

### LINKED ISSUE #817: [Bug]: Figure out why `TREFHT` regridded reference variable is not regridding on `main` with ESMF bilinear
### What happened?

In PR #815:

> This PR validates the replication of the slice flag feature added in PR #750 to fix #759.
> 
> I re-ran `ex1.py` and found that all `.nc` files produced are now within the relative tolerance.
> 
> For metrics, all but one is within the relative tolerance. This single metric is related to the regridded reference data for `"TREFHT"` with the `land` region. In [trefht.py](https://github.com/E3SM-Project/e3sm_diags/pull/815/files#diff-35f583b4eda8fd8effb76e44533f3d7b4daa5f6b80499a4bb0f8c8e2ef1b06a9), I compare the `.nc` file for the regridded reference data between xCDAT/xESMF and CDAT/ESMF. I found that CDAT/ESMF doesn't seem to actually regrid the data since the original and regridded reference variables are the same, as noted in this PR review [comment](https://github.com/E3SM-Project/e3sm_diags/pull/815#discussion_r1625034843). I think this is leading to the large diffs along the coast lines, as shown below.

| xCDAT         | CDAT     | Diff |
|--------------|-----------|------------|
|    debug_TREFHT_actual | debug_TREFHT_expected     |     debug_TREFHT_diff

> In PR #794, I noticed how bilinear regridding with xCDAT/xESMF produces nan location mismatches when comparing land sea masks to CDAT/ESMF. I think we should move forward for now and debug this more later on in a separate PR.

### What did you expect to happen? Are there are possible answers you came across?

The reference `TREFHT` variable should be regridded since the `mv2.regrid` is called below:

https://github.com/E3SM-Project/e3sm_diags/blob/c8404ab0f5523fee80cda396357f307b7ad1ff5b/e3sm_diags/driver/utils/general.py#L246-L254

### Minimal Complete Verifiable Example (MVCE)

_No response_

### Relevant log output

_No response_

### Anything else we need to know?

_No response_

### Environment

`main` branch

### LINKED ISSUE #825: [Bug]: `qbo` set `unify_plev()` does not return new variable with sorted Z axis
### What happened?

This bug was caught while working on #664. `unify_plev()` is supposed to sort the Z axis in ascending order. However, this operation is not performed in-place/in-memory on the object so the Z axis is never updated.  

https://github.com/E3SM-Project/e3sm_diags/blob/898329fa1e55279944df4f59e1598a68d5bb8e13/e3sm_diags/driver/qbo_driver.py#L180-L182

https://github.com/E3SM-Project/e3sm_diags/blob/898329fa1e55279944df4f59e1598a68d5bb8e13/e3sm_diags/driver/qbo_driver.py#L22-L35

I'm not sure if this bug has been affecting any downstream computations or plotting.

### What did you expect to happen? Are there are possible answers you came across?

1. `unify_plev()` needs to return the `var` object
2. Update qbo driver `test_region = unify_plev(test_region)` and `ref_region = unify_plev(ref_region)`

### Minimal Complete Verifiable Example (MVCE)

_No response_

### Relevant log output

_No response_

### Anything else we need to know?

_No response_

### Environment

Latest `e3sm_diags` on `main` and stable build

### LINKED ISSUE #854: [Bug]: `model_vs_model` `lat_lon` variables not generated on `main` but generated on `cdat-migration-fy24`
### What happened?

In PR #846, the dev branch is able to generate outputs for the variables below. However, they are missing on `main` which means `main` might not be able to generate them for some reason.

```python
['/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-OMEGA-200-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-OMEGA-200-JJA-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-OMEGA-500-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-OMEGA-500-JJA-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-OMEGA-850-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-OMEGA-850-JJA-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-T-200-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-T-200-JJA-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-T-850-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-T-850-JJA-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-U-200-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-U-200-JJA-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-U-850-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-U-850-JJA-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-Z3-500-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/cdat-migration-fy24/main-model-vs-model/lat_lon/model_vs_model/-Z3-500-JJA-global_test.nc']
```

### What did you expect to happen? Are there are possible answers you came across?

_No response_

### Minimal Complete Verifiable Example (MVCE)

_No response_

### Relevant log output

_No response_

### Anything else we need to know?

_No response_

### Environment

Latest `main`

### LINKED ISSUE #899: [Bug]: dependence on setuptools needed
### What happened?

We need to add setuptools as a build-time dependency, preferably with some kind of constraint.  pip no longer includes setuptools by default for python >=3.13.

### What did you expect to happen? Are there are possible answers you came across?

_No response_

### Minimal Complete Verifiable Example (MVCE)

_No response_

### Relevant log output

_No response_

### Anything else we need to know?

_No response_

### Environment

N/A

See https://github.com/conda-forge/e3sm_diags-feedstock/pull/38

### PR DESCRIPTION
**Pull request by tomvothecoder**
### Misc. issues
- Closes #899

### Bugs on `main` that are addressed on this branch 
- Closes #825
- Closes #854
- Closes #817
- Closes #797

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

If applicable:

- [x] New and existing unit tests pass with my changes (locally and CI/CD build)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: docs/source/available-parameters.rst [hunk 1]
```diff
@@ -67,15 +67,19 @@ functionality of the diagnostics.
    See `default_regions.py
    <https://github.com/E3SM-Project/e3sm_diags/blob/master/e3sm_diags/derivations/default_regions.py>`__
    for a list of possible regions. Ex: ``regions=["global", "TROPICS"]``.
--  **regrid_method**: What regird method of the regrid tool to use.
+-  **regrid_method**: What grid method of the regrid tool to use.
```
Comment (tomvothecoder) [inline docs/source/available-parameters.rst:70]:
```suggestion
-  **regrid_method**: What regrid method of the regrid tool to use.
```

Comment (tomvothecoder) [inline docs/source/available-parameters.rst:70]:
```suggestion
-  **regrid_method**: What regrid method of the regrid tool to use.
```

File: e3sm_diags/derivations/default_regions_xr.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
+"""Module for defining regions used for spatial subsetting.
+
+NOTE: Replaces `e3sm_diags.derivations.default_regions`.
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/default_regions_xr.py:3]:
```suggestion
```

File: e3sm_diags/driver/utils/climo_xr.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
+"""This module stores climatology functions operating on Xarray objects.
+
+NOTE: Replaces `e3sm_diags.driver.utils.climo`.
```
Comment (tomvothecoder) [inline e3sm_diags/driver/utils/climo_xr.py:3]:
```suggestion
```

File: e3sm_diags/driver/utils/dataset_xr.py [hunk 4]
```diff
@@ -0,0 +1,1488 @@
+"""This module stores the Dataset class, which is the primary class for I/O.
+
+NOTE: Replaces `e3sm_diags.driver.utils.dataset`.
```
Comment (tomvothecoder) [inline e3sm_diags/driver/utils/dataset_xr.py:3]:
```suggestion
```

File: e3sm_diags/metrics/metrics.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+    Returns
+    -------
+    xr.DataArray
+        Weights for the specified axis.
+
+    Notes
+    -----
+    xCDAT's ``ds.spatial.get_weights()`` method only supports rectilinear grids
+    # ("X", "Y") as of v0.6.1. This function computes weights for "X" and Y"
```
Comment (tomvothecoder) [inline e3sm_diags/metrics/metrics.py:213]:
```suggestion
    ("X", "Y") as of v0.6.1. This function computes weights for "X" and Y"
```

**Review by tomvothecoder — COMMENTED** [review]
Misc. code changes from self-review

**Comment by tomvothecoder** [comment]
@chengzhuzhang I plan to merge this either today or tomorrow

**Comment by chengzhuzhang** [comment]
@tomvothecoder thanks for the heads-up. Please charge forward!

File: e3sm_diags/derivations/default_regions_xr.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
+"""Module for defining regions used for spatial subsetting.
+
+"""
```
Comment (tomvothecoder) [inline e3sm_diags/derivations/default_regions_xr.py:3]:
```suggestion
"""Module for defining regions used for spatial subsetting."""
```

**Pull request merged**